src/rgw/rgw_lc.cc

```diff
@@ -539,6 +539,36 @@ static bool pass_size_limit_checks(const DoutPrefixProvider *dpp, lc_op_ctx& oc)
 static std::string lc_id = "rgw lifecycle";
 static std::string lc_req_id = "0";
 
+static void send_notification(const DoutPrefixProvider *dpp,
+                              rgw::sal::Driver *driver,
+                              rgw::sal::Object *obj,
+                              rgw::sal::Bucket *bucket,
+                              const std::string &etag,
+                              const std::string &version_id,
+                              const rgw::notify::EventTypeList &event_types) {
+// notification supported only for RADOS driver for now
+  auto notify = driver->get_notification(
+      dpp, obj, nullptr, event_types, bucket, lc_id,
+      const_cast<std::string &>(bucket->get_tenant()), lc_req_id, null_yield);
+
+  int ret = notify->publish_reserve(dpp, nullptr);
+  if (ret < 0) {
+    ldpp_dout(dpp, 1) << "ERROR: notify publish_reserve failed, with error: "
+                      << ret << " for lc object: " << obj->get_name()
+                      << " for event_types: " << event_types
+                      << dendl;
+    return;
+  }
+  ret =
+      notify->publish_commit(dpp, obj->get_obj_size(), ceph::real_clock::now(),
```

Comment:
Probably this will get replaced with new SAL APIs to fetch right obj_state & size from the backend store. 
 
Are there any test-cases to check notifications with LC expiration & cloud-transition?  Post transition to cloud, either the object's HEAD is retained with data erased (so the `size` will be '0' bytes) or it is deleted (same as the case of LC expiration), in which in case get_obj_size() may return ENOENT if that call is passed to backend store.

Request @cbodley @dang to comment.

Reply:
I have not made any changes as part of this PR to use sal object, I have just reorganised the code
If a future PR is gonna change this behaviour then should this PR be blocked for that ?

Reply:
cc @yuvalif 

the bucket notification tests do include some coverage for lifecycle, but not cloud transition:
https://github.com/ceph/ceph/blob/a052683/src/test/rgw/bucket_notification/test_bn.py#L1720-L1913

Reply:
> in which in case get_obj_size() may return ENOENT if that call is passed to backend store.

similar to `Object::get_attrs()`, [get_obj_size()](https://github.com/ceph/ceph/blob/a052683/src/rgw/rgw_sal_store.h#L232) is just returning the cached object state, so shouldn't ever go to the backend. some mutations may update that cached state, but i don't think `delete_obj()` does. callers probably shouldn't expect getters like `get_obj_size()` to return sensible answers after deletion

Reply:
> > in which in case get_obj_size() may return ENOENT if that call is passed to backend store.
> 
> similar to `Object::get_attrs()`, [get_obj_size()](https://github.com/ceph/ceph/blob/a052683/src/rgw/rgw_sal_store.h#L232) is just returning the cached object state, so shouldn't ever go to the backend. some mutations may update that cached state, but i don't think `delete_obj()` does. callers probably shouldn't expect getters like `get_obj_size()` to return sensible answers after deletion

@cbodley so you suggesting, instead of calling the `get_obj_size` after the object is removed (lc'd), just get the size prior of deletion ?

Reply:
like was done for etag, yeah